src/serde/de.rs

```diff
@@ -0,0 +1,997 @@
+                deserializer: self,
+            });
+            self.reader.next()?;
+            return result;
+        }
+
+        self.reader.step_in()?;
+        self.reader.next()?;
+        let result = visitor.visit_map(&mut *self)?;
+        self.reader.next()?;
+        Ok(result)
+    }
+
+    fn deserialize_enum<V>(
+        self,
+        _name: &'static str,
+        _variants: &'static [&'static str],
+        visitor: V,
+    ) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        if let Some(ion_type) = self.reader.ion_type() {
+            match ion_type {
+                IonType::String => visitor.visit_enum(UnitVariantAccess::new(self)),
+                IonType::Struct => {
+                    self.reader.step_in()?;
+                    self.reader.next()?;
+                    let value = visitor.visit_enum(VariantAccess::new(self))?;
+                    self.reader.step_out()?;
+                    self.reader.next()?;
+                    Ok(value)
+                }
+                _ => Err(IonError::DecodingError {
+                    description: "expeted an enumeration".to_string(),
+                }),
+            }
+        } else {
+            Err(IonError::DecodingError {
+                description: "unexpected end of file".to_string(),
+            })
+        }
+    }
+
+    fn deserialize_identifier<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        let result = visitor.visit_string(self.reader.read_string()?);
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_ignored_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        self.deserialize_any(visitor)
+    }
+}
+
+impl<'de, 'a, R> SeqAccess<'de> for &'a mut Deserializer<R>
+where
+    R: StreamReader<Symbol = Symbol, Item = StreamItem>,
+{
+    type Error = IonError;
+
+    fn next_element_seed<T>(&mut self, seed: T) -> Result<Option<T::Value>, Self::Error>
+    where
+        T: DeserializeSeed<'de>,
+    {
+        if self.reader.ion_type().is_none() {
+            if self.reader.depth() > 0 {
+                self.reader.step_out()?;
+            }
+            Ok(None)
+        } else {
+            seed.deserialize(&mut **self).map(Some)
+        }
+    }
+}
+
+impl<'de, 'a, R> MapAccess<'de> for &'a mut Deserializer<R>
+where
+    R: StreamReader<Symbol = Symbol, Item = StreamItem>,
+{
+    type Error = IonError;
+
+    fn next_key_seed<K>(&mut self, seed: K) -> Result<Option<K::Value>, Self::Error>
+    where
+        K: DeserializeSeed<'de>,
+    {
+        if self.reader.ion_type().is_none() {
+            self.reader.step_out()?;
+            Ok(None)
+        } else {
+            let key = self.reader.field_name().map(|x| x.to_string())?;
+            let deserializer = MapKeyDeserializer { key };
+            Ok(Some(seed.deserialize(deserializer)?))
```

Comment:
I'm not sure what `MapKeyDeserializer` does, but is copying every struct name into its own string necessary?